nixos/modules/system/etc/setup-etc.pl

```diff
@@ -1,82 +1,111 @@
-    return 1;
+    unlink($tmp);
+    if (-e $tmp) {
+        die("Failed to remove `$tmp` file before linking");
+    }
+    symlink($source, $tmp) or die("Failed to create symlink `$tmp` -> `$source`: $!");
+    rename($tmp, $target) or die("Failed to move `$tmp` -> `$target`: $!");
 }
 
 
 # Atomically update /etc/static to point at the etc files of the
 # current configuration.
-atomicSymlink $etc, $static or die;
+try {
+    atomic_symlink($etc, $static);
+}
+catch ($e) {
+    die("Failed to create symlink for /etc: $e");
+}
+
 
 # Returns 1 if the argument points to the files in /etc/static.  That
 # means either argument is a symlink to a file in /etc/static or a
 # directory with all children being static.
-sub isStatic {
-    my $path = shift;
-
+sub is_static ($path) {
     if (-l $path) {
-        my $target = readlink $path;
-        return substr($target, 0, length "/etc/static/") eq "/etc/static/";
+        my $target = readlink($path);
+        return substr($target, 0, length($static)) eq $static;
     }
 
     if (-d $path) {
-        opendir DIR, "$path" or return 0;
-        my @names = readdir DIR or die;
-        closedir DIR;
-
+        my @names = read_dir($path);
         foreach my $name (@names) {
-            next if $name eq "." || $name eq "..";
-            unless (isStatic("$path/$name")) {
-                return 0;
+            if (not (is_static("$path/$name"))) {
+                return;
             }
         }
         return 1;
     }
-
-    return 0;
+    return;
 }
 
 # Remove dangling symlinks that point to /etc/static.  These are
 # configuration files that existed in a previous configuration but not
 # in the current one.  For efficiency, don't look under /etc/nixos
-# (where all the NixOS sources live).
+# (where all the NixOS sources live) and /etc/static which is what we're
+# comparing to anyway.
 sub cleanup {
-    if ($File::Find::name eq "/etc/nixos") {
-        $File::Find::prune = 1;
-        return;
-    }
-    if (-l $_) {
-        my $target = readlink $_;
-        if (substr($target, 0, length $static) eq $static) {
-            my $x = "/etc/static/" . substr($File::Find::name, length "/etc/");
-            unless (-l $x) {
-                print STDERR "removing obsolete symlink ‘$File::Find::name’...\n";
-                unlink "$_";
+    try {
+        my $filename = $_;
+        if ($File::Find::name eq "/etc/nixos" || $File::Find::name eq $static) {
+            $File::Find::prune = 1;
+            return;
+        }
+        if (-l $filename) {
+            my $target = readlink($filename);
+            if (substr($target, 0, length($static)) eq $static) {
+                my $x = $static . substr($File::Find::name, length("/etc/"));
+                # Check if symlink is still present in new etc
+                if (not (-l $x)) {
+                    warn("removing obsolete symlink ‘$File::Find::name’...\n");
+                    unlink($_) or die("Failed to remove `$_`: $!");
```

Comment:
Good catch!

BTW: You can also avoid the `my $filename = $_;` by using the signature feature when defining the subroutine.

```perl
sub cleanup ($filename) {
```

Similarly in `make_symlinks`

Reply:
Unfortunately, in this case the signature doesn't work as `find()` won't supply an argument to the function but rather set the `$_` variable https://perldoc.perl.org/File::Find#$_-is-the-current-filename-within-that-directory
Otherwise this would've worked i think 🤔

Reply:
Same goes for the `make_symlinks` function

Reply:
Yeah, sorry, you're totally right on that I think. The code does just set `$_`, passing no arguments, it seems. Setting `my $filename = $_;` and using that is the best we can do then it seems.

https://metacpan.org/release/RJBS/perl-5.36.0/source/ext/File-Find/lib/File/Find.pm#L758-761
```perl
sub find {
    my $wanted = shift;
    _find_opt(wrap_wanted($wanted), @_);
}
```
https://metacpan.org/release/RJBS/perl-5.36.0/source/ext/File-Find/lib/File/Find.pm#L713-756
```perl
sub wrap_wanted {
    my $wanted = shift;
    if ( ref($wanted) eq 'HASH' ) {
        # -- snip --
    }
    elsif( ref( $wanted ) eq 'CODE' ) {
        return { wanted => $wanted };
    }
    else {
       die 'no &wanted subroutine given';
    }
}
```
https://metacpan.org/release/RJBS/perl-5.36.0/source/ext/File-Find/lib/File/Find.pm#L126-285
```perl
sub _find_opt {
    my $wanted = shift;
# --snip --
    $wanted_callback   = $wanted->{wanted};
# --snip --
            $_ = $name if $no_chdir;
 
            { $wanted_callback->() }; # protect against wild "next"
# --snip --
}
```